src/encoder/serialization/serde.rs

```diff
@@ -0,0 +1,129 @@
+use std::collections::HashMap;
+use serde;
+use {StrVal, Bool, VecVal, Map, OptVal};
+use encoder::{Encoder, EncoderResult, Error};
+use encoder::Error::*;
+
+pub use serde::Serialize as Encodable;
+
+impl serde::ser::Error for Error {
+    fn custom<T: Into<String>>(msg: T) -> Self {
+        Error::Serialization(msg.into())
+    }
+}
+
+impl serde::Serializer for Encoder {
+    type Error = Error;
+
+    fn serialize_bool(&mut self, v: bool) -> EncoderResult {
+        self.push_ok(Bool(v))
+    }
+
+    fn serialize_i64(&mut self, v: i64) -> EncoderResult {
+        self.push_str_ok(v)
+    }
+
+    fn serialize_u64(&mut self, v: u64) -> EncoderResult {
+        self.push_str_ok(v)
+    }
+
+    fn serialize_f64(&mut self, v: f64) -> EncoderResult {
+        self.push_str_ok(v)
+    }
+
+    fn serialize_str(&mut self, v: &str) -> EncoderResult {
+        self.push_str_ok(v)
+    }
+
+    fn serialize_unit(&mut self) -> EncoderResult {
+        Err(UnsupportedType)
+    }
+
+    fn serialize_none(&mut self) -> EncoderResult {
+        self.push_ok(OptVal(None))
+    }
+
+    fn serialize_some<V>(&mut self, value: V) -> EncoderResult
+    where V: serde::ser::Serialize
+    {
+        try!(value.serialize(self));
+
+        let val = match self.data.pop() {
+            Some(OptVal(_)) => return Err(NestedOptions),
+            Some(d) => d,
+            _ => {
+                return Err(UnsupportedType);
+            }
+        };
+        self.push_ok(OptVal(Some(Box::new(val))))
+    }
+
+    fn serialize_seq<V>(&mut self, mut visitor: V) -> EncoderResult
```

Comment:
The way `Serializer` deals with seqs and maps is changing (in master, but not crates.io). Check out: https://github.com/serde-rs/serde/pull/437

Instead of using visitors, you have a single associated type for handling the state of a map being serialised.